compiler/src/dotty/tools/dotc/interactive/Completion.scala

```diff
@@ -410,23 +412,34 @@ object Completion {
     val Import: Mode = new Mode(4) | Term | Type
   }
 
-  /** A scope that tracks renames of the entered symbols.
-   *  Useful for providing completions for renamed symbols
-   *  in the REPL and the IDE.
-   */
-  private class RenameAwareScope extends Scopes.MutableScope {
-    private val nameToSymbols: mutable.Map[TermName, List[Symbol]] = mutable.Map.empty
-
-    /** Enter the symbol `sym` in this scope, recording a potential renaming. */
-    def enter[T <: Symbol](sym: T, name: Name)(using Context): T = {
-      val termName = name.stripModuleClassSuffix.toTermName
-      nameToSymbols += termName -> (sym :: nameToSymbols.getOrElse(termName, Nil))
-      newScopeEntry(name, sym)
-      sym
-    }
+  private object CompletionScope  {
+    val empty = CompletionScope()
+  }
+
+  private case class CompletionScope(nameToDenots: Map[Name, List[SingleDenotation]] = Map.empty) {
+    def mergeShadowedBy(that: CompletionScope) = CompletionScope(this.nameToDenots ++ that.nameToDenots)
+
+    def mergeDiscardingAmbiguities(that: CompletionScope) =
+      val mappings = (this.nameToDenots.toList ++ that.nameToDenots.toList)
+        .groupBy(_._1)
+        .collect {
+          case (name, (_, denots) :: Nil)  => name -> denots
+        }
+      CompletionScope(mappings)
 
-    /** Get the names that are known in this scope, along with the list of symbols they refer to. */
-    def mappings: Map[TermName, List[Symbol]] = nameToSymbols.toMap
+    /**
+     * Return the list of symbols that should be included in completion results.
+     *
+     * If several symbols share the same name, the type symbols appear before term symbols inside
+     * the same `Completion`.
+     */
+    def getCompletions(using Context): List[Completion] = {
+      nameToDenots.toList.groupBy(_._1.toTermName.show).map { (name, namedDenots) =>
```

Comment:
Can we avoided the `groupBy` here?

Reply:
I don't think so if we want to keep the current semantics although I'm still wondering of this is what we want for the future https://github.com/lampepfl/dotty/issues/11313 but changing this is not an intension of this PR